module/x/peggy/abci.go

```diff
@@ -0,0 +1,141 @@
+package peggy
+
+import (
+	"github.com/althea-net/peggy/module/x/peggy/keeper"
+	"github.com/althea-net/peggy/module/x/peggy/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	stakingtypes "github.com/cosmos/cosmos-sdk/x/staking/types"
+)
+
+// EndBlocker is called at the end of every block
+func EndBlocker(ctx sdk.Context, k keeper.Keeper) {
+	params := k.GetParams(ctx)
+	currentBondedSet := k.StakingKeeper.GetBondedValidatorsByPower(ctx)
+
+	// valsets are sorted so the most recent one is first
+	valsets := k.GetValsets(ctx)
+	if len(valsets) == 0 {
+		k.SetValsetRequest(ctx)
+	}
+
+	for i, vs := range valsets {
+		switch {
+		// #1 condition
+		// We look through the full bonded validator set (not just the active set, include unbonding validators)
+		// and we slash users who haven't signed a valset that is currentHeight - signedBlocksWindow old
+		case uint64(ctx.BlockHeight())-params.SignedValsetsWindow > vs.Height:
+
+			// first we need to see which validators in the active set
+			// haven't signed the valdiator set and slash them,
+			var toSlash []stakingtypes.Validator
+			confirms := k.GetValsetConfirms(ctx, vs.Nonce)
+			for _, val := range currentBondedSet {
+				found := false
+				for _, conf := range confirms {
+					if conf.EthAddress == k.GetEthAddress(ctx, val.GetOperator()) {
+						found = true
+						break
+					}
+				}
+				if !found {
+					toSlash = append(toSlash, val)
+				}
+			}
+
+			for _, val := range toSlash {
+				cons, _ := val.GetConsAddr()
+				k.StakingKeeper.Slash(ctx, cons, ctx.BlockHeight(), val.ConsensusPower(), params.SlashFractionValset)
+				k.StakingKeeper.Jail(ctx, cons)
+			}
+
+			// then we prune the valset from state
+			k.DeleteValset(ctx, vs.Nonce)
+
+		// on the latest validator set, check for change in power against
+		// current, and emit a new validator set if the change in power >1%
+		case i == 0:
+			if types.BridgeValidators(k.GetCurrentValset(ctx).Members).PowerDiff(vs.Members) > 0.01 {
+				k.SetValsetRequest(ctx)
+			}
+		}
+	}
+
+	// #2 condition
+	// We look through the full bonded set (not just the active set, include unbonding validators)
+	// and we slash users who haven't signed a batch confirmation that is >15hrs in blocks old
+	batches := k.GetOutgoingTxBatches(ctx)
+	for _, batch := range batches {
+		if uint64(ctx.BlockHeight())-params.SignedBatchesWindow > batch.Block {
```

Comment:
super nit, but I find the code this more readable when you move this to a const:

```go
signedWithinWindow := uint64(ctx.BlockHeight())- batch.Block > params.SignedBatchesWindow
if signedWithinWindow {
// ...
}
```